### PR TITLE
Track peer traffic by category

### PR DESCRIPTION
**Pull request by JoelKatz**
Report peer traffic, both inbound and outbound, by category.

### CODE REVIEW

File: src/ripple/overlay/impl/TrafficCount.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        std::map <std::string,
+            std::pair <unsigned long, unsigned long>>
+                ret;
+
+        for (auto& i : _counts)
+        {
+            ret.emplace (getName(i.first),
+                std::pair <unsigned long, unsigned long>
+                    (i.second.first.load(),
+                    i.second.second.load()));
```
Comment (vinniefalco) [inline src/ripple/overlay/impl/TrafficCount.h]: Alternative:
```
ret.emplace(std::piecewise_construct,
  std::make_tuple(getName(i.first)),
    std::make_tuple(i.second.first.load(), i.second.second.load()));
```

File: src/ripple/overlay/impl/TrafficCount.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        CT_get_ledger,     // ledgers we try to get
+        CT_share_ledger,   // ledgers we share
+        CT_get_trans,      // transaction sets we try to get
+        CT_share_trans,    // transaction sets we get
+        CT_unknown         // must be last
+    };
+
+    static const char* getName (Category c);
+
+    static Category categorize (
```
Comment (vinniefalco) [inline src/ripple/overlay/impl/TrafficCount.h]: In theory this should be named `category` but far from a dealbreaker

File: src/ripple/overlay/impl/TrafficCount.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                    (i.second.first.load(),
+                    i.second.second.load()));
+        }
+
+        return ret;
+    }
+
+    protected:
+
+    std::map <Category, Count_t> _counts;
```
Comment (vinniefalco) [inline src/ripple/overlay/impl/TrafficCount.h]: `counts_` but not worth holding up for

**Comment by vinniefalco** [comment]
Missing `#include <ripple/overlay/impl/TrafficCount.h>` in Message.cpp

**Comment by vinniefalco** [comment]
:+1:

File: src/ripple/overlay/impl/OverlayImpl.cpp [hunk 4]
```diff
@@ -588,6 +589,20 @@ OverlayImpl::onChildrenStopped ()
...
+    for (auto& i : stats)
+    {
+        if (! i.second.first && ! i.second.second)
+            continue;
+
+        beast::PropertyStream::Map item (set);
+        item["category"] = i.first;
+        item["bytes_in"] =
+            beast::lexicalCast<std::string>(i.second.first);
+        item["bytes_out"] =
```
Comment (vinniefalco) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: Why don't we report the number of messages in addition to the total bytes?

**Comment by vinniefalco** [comment]
Its not critical, but I would have made the enumeration of message types a public declaration in `Message.h` so that it could stay in its native type instead of being casted to `int`.

File: src/ripple/overlay/impl/OverlayImpl.cpp (hunk 4 above)
Comment (JoelKatz) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: I can add that easily enough.

**Comment by JoelKatz** [comment]
It was difficult to keep the native type in Message before the TrafficCount was moved to impl. I meant to do that when I moved to impl but it got forgotten.

**Comment by vinniefalco** [comment]
@JoelKatz I'm not losing sleep over it, but it highlights how annoying protocol buffers are that you have to store the message integer yourself, instead of it being a part of the message.

**Comment by nbougalis** [comment]
:+1:

**Comment by nbougalis** [comment]
Merged as 61e5359231b304bd7c950c8e17e7092872df7737.

**Pull request closed without merging**